packages/nouns-contracts/contracts/NounsDescriptorV2.sol

```diff
@@ -0,0 +1,427 @@
+     */
+    function setBaseURI(string calldata _baseURI) external override onlyOwner {
+        baseURI = _baseURI;
+
+        emit BaseURIUpdated(_baseURI);
+    }
+
+    /**
+     * @notice Given a token ID and seed, construct a token URI for an official Nouns DAO noun.
+     * @dev The returned value may be a base64 encoded data URI or an API URL.
+     */
+    function tokenURI(uint256 tokenId, INounsSeeder.Seed memory seed) external view override returns (string memory) {
+        if (isDataURIEnabled) {
+            return dataURI(tokenId, seed);
+        }
+        return string(abi.encodePacked(baseURI, tokenId.toString()));
+    }
+
+    /**
+     * @notice Given a token ID and seed, construct a base64 encoded data URI for an official Nouns DAO noun.
+     */
+    function dataURI(uint256 tokenId, INounsSeeder.Seed memory seed) public view override returns (string memory) {
+        string memory nounId = tokenId.toString();
+        string memory name = string(abi.encodePacked('Noun ', nounId));
+        string memory description = string(abi.encodePacked('Noun ', nounId, ' is a member of the Nouns DAO'));
+
+        return genericDataURI(name, description, seed);
+    }
+
+    /**
+     * @notice Given a name, description, and seed, construct a base64 encoded data URI.
+     */
+    function genericDataURI(
+        string memory name,
+        string memory description,
+        INounsSeeder.Seed memory seed
+    ) public view override returns (string memory) {
+        NFTDescriptorV2.TokenURIParams memory params = NFTDescriptorV2.TokenURIParams({
+            name: name,
+            description: description,
+            parts: getPartsForSeed(seed),
+            background: art.backgrounds(seed.background)
+        });
+        return NFTDescriptorV2.constructTokenURI(renderer, params);
+    }
+
+    /**
+     * @notice Given a seed, construct a base64 encoded SVG image.
+     */
+    function generateSVGImage(INounsSeeder.Seed memory seed) external view override returns (string memory) {
+        ISVGRenderer.SVGParams memory params = ISVGRenderer.SVGParams({
+            parts: getPartsForSeed(seed),
+            background: art.backgrounds(seed.background)
+        });
+        return NFTDescriptorV2.generateSVGImage(renderer, params);
+    }
+
+    /**
+     * @notice Get all Noun parts for the passed `seed`.
+     */
+    function getPartsForSeed(INounsSeeder.Seed memory seed) public view returns (ISVGRenderer.Part[] memory) {
+        bytes memory _body = art.bodies(seed.body);
+        bytes memory _accessory = art.accessories(seed.accessory);
+        bytes memory _head = art.heads(seed.head);
+        bytes memory _glassesBytes = art.glasses(seed.glasses);
+
+        ISVGRenderer.Part[] memory _parts = new ISVGRenderer.Part[](4);
+        _parts[0] = ISVGRenderer.Part({ image: _body, palette: _getPalette(_body) });
+        _parts[1] = ISVGRenderer.Part({ image: _accessory, palette: _getPalette(_accessory) });
```

Comment:
there's room for some optimization here when calling `_getPalette`, we could cache the values instead of calling `art.palettes`, but maybe not worth it